src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -369,210 +379,114 @@ public ExponentialHistogramData GetExponentialHistogramData()
-                }
+        this.CompleteUpdate();
+    }
 
-            case AggregationType.HistogramWithMinMaxBuckets:
-                {
-                    this.UpdateHistogramWithBucketsAndMinMax((double)number);
-                    break;
-                }
+    internal void Update(double number, ReadOnlySpan<KeyValuePair<string, object?>> tags)
+    {
+        this.Update(number, out var explicitBucketHistogramBucketIndex);
 
-            case AggregationType.Base2ExponentialHistogram:
-                {
-                    this.UpdateBase2ExponentialHistogram((double)number);
-                    break;
-                }
+        var exemplarFilter = this.aggregatorStore.ExemplarFilter;
+        if (exemplarFilter.EarlySampleDecision ?? exemplarFilter.ShouldSampleDouble(number, tags))
+        {
+            Debug.Assert(this.mpComponents?.ExemplarReservoir != null, "ExemplarReservoir was null");
 
-            case AggregationType.Base2ExponentialHistogramWithMinMax:
-                {
-                    this.UpdateBase2ExponentialHistogramWithMinMax((double)number);
-                    break;
-                }
+            this.mpComponents!.ExemplarReservoir!.Offer(
+                new ExemplarMeasurement<double>(number, tags, explicitBucketHistogramBucketIndex));
         }
 
-        // There is a race with Snapshot:
-        // Update() updates the value
-        // Snapshot snapshots the value
-        // Snapshot sets status to NoCollectPending
-        // Update sets status to CollectPending -- this is not right as the Snapshot
-        // already included the updated value.
-        // In the absence of any new Update call until next Snapshot,
-        // this results in exporting an Update even though
-        // it had no update.
-        // TODO: For Delta, this can be mitigated
-        // by ignoring Zero points
-        this.MetricPointStatus = MetricPointStatus.CollectPending;
+        this.CompleteUpdate();
+    }
 
-        if (this.aggregatorStore.OutputDeltaWithUnusedMetricPointReclaimEnabled)
+    internal void TakeSnapshot(bool outputDelta)
+    {
+        this.Snapshot(outputDelta);
+
+        var exemplarReservoir = this.mpComponents?.ExemplarReservoir;
+        if (exemplarReservoir != null)
         {
-            Interlocked.Decrement(ref this.ReferenceCount);
+            this.mpComponents!.Exemplars = this.mpComponents.ExemplarReservoir!.Collect();
         }
     }
 
-    internal void UpdateWithExemplar(long number, ReadOnlySpan<KeyValuePair<string, object?>> tags, bool isSampled)
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void Update(long number, out int explicitBucketHistogramBucketIndex)
     {
-        Debug.Assert(this.mpComponents != null, "this.mpComponents was null");
-
         switch (this.aggType)
         {
             case AggregationType.LongSumIncomingDelta:
                 {
-                    this.mpComponents!.AcquireLock();
```

Comment:
The current code was using a lock() (implemented with CAS calls/spin-wait, not lock() statement.), and then doing everything inside that lock - updating the sum, doing the exemplar storage.

Now we don't do that lock, but uses **multiple** interlocked calls - one to update the sum, one to update measurementsSeen, one (or possibly many) more to get the right buffer, and one more to make sure no one else is updating simultaneously.

Despite this, the perf gains in stress test is significant. What could be the a good reasoning for this?

Reply:
There is just a lot less contention/spinning going on.

The stress test runs 20 threads on my machine all hammering measurements to the same counter. The measurements are writing the same tag "DimName1" with one of ten random values.

Before this PR if two measurements happen on the same random value, one of those threads is going to have to spin for the duration of the value update + exemplar update. Exemplar update can be very slow. Best case is you wait for a random value and then you don't record. Worst case is you wait for a random value and then you record which is a bunch of writes.

After this PR all measurements do an Interlocked operation for the value record portion so that spin is eliminated. Then they enter exemplar world and get a thread-local random. If there isn't going to be a record the thread is done. If there is a record, we now have `Environment.ProcessorCount` number of buckets available. Each thread may write to one of those buckets without blocking anything. If some thread happens to be writing to the same bucket, we just abort the record (we don't spin). Essentially in this design there is no spinning for a measurement so the threads spend more time doing work and we get better throughput.

Reply:
I ran stress tests against the latest version of this PR.

* Counters are improved in all cases.
* Histogram
   * Improved when there is contention (lock is held for less time).
   * Slightly degraded when there isn't a lot of contention. This is due to a few extra operations being executed now in the exemplar path. I could special-case something to fix this. But if you look at the overall throughput of histograms vs counters, they suck 🤣 I would like to look at that on a follow-up and see if I can use the same active/passive buffer tricks to help histograms across the board.

## Counters

New version using counter w/ tag-filtering view + exemplars:

* Total Runway Time (seconds) 300
* Total Loops: 9,899,082,415
* Average Loops/Second: 32,964,636
* Average CPU Cycles/Loop: 1,727
* GC Total Allocated Bytes: 10643640

Old version using counter w/ tag-filtering view + exemplars:

* Total Runway Time (seconds) 300
* Total Loops: 5,076,211,789
* Average Loops/Second: 16,909,264
* Average CPU Cycles/Loop: 869
* GC Total Allocated Bytes: 11098120

New version using counter w/ exemplars:

* Total Runway Time (seconds) 300
* Total Loops: 10,881,621,737
* Average Loops/Second: 36,250,443
* Average CPU Cycles/Loop: 1,573
* GC Total Allocated Bytes: 14448912

Old version using counter w/ exemplars:

* Total Runway Time (seconds) 300
* Total Loops: 8,958,303,265
* Average Loops/Second: 29,855,139
* Average CPU Cycles/Loop: 1,855
* GC Total Allocated Bytes: 11546904

## Histograms

New version using histogram w/ tag-filtering view + exemplars:

* Total Runway Time (seconds) 300
* Total Loops: 3,069,842,084
* Average Loops/Second: 10,227,489
* Average CPU Cycles/Loop: 3,781
* GC Total Allocated Bytes: 11073856

Old version using histogram w/ tag-filtering view + exemplars:

* Total Runway Time (seconds) 300
* Total Loops: 2,160,807,746
* Average Loops/Second: 7,197,558
* Average CPU Cycles/Loop: 2,160
* GC Total Allocated Bytes: 16985496

New version using histogram w/ exemplars:

* Total Runway Time (seconds) 300
* Total Loops: 5,576,901,325
* Average Loops/Second: 18,583,167
* Average CPU Cycles/Loop: 3,001
* GC Total Allocated Bytes: 177120704

Old version using histogram w/ exemplars:

* Total Runway Time (seconds) 300
* Total Loops: 5,767,724,875
* Average Loops/Second: 19,220,944
* Average CPU Cycles/Loop: 2,733
* GC Total Allocated Bytes: 265700816